xds/src/main/java/io/grpc/xds/internal/certprovider/MeshCaCertificateProviderProvider.java

```diff
@@ -0,0 +1,207 @@
+
+import com.google.common.annotations.VisibleForTesting;
+import io.grpc.internal.BackoffPolicy;
+import io.grpc.internal.ExponentialBackoffPolicy;
+import io.grpc.xds.internal.sts.StsCredentials;
+import java.util.Map;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Provider of {@link CertificateProvider}s. Implemented by the implementer of the plugin. We may
+ * move this out of the internal package and make this an official API in the future.
+ */
+final class MeshCaCertificateProviderProvider implements CertificateProviderProvider {
+
+  private static final Logger logger = Logger
+      .getLogger(MeshCaCertificateProviderProvider.class.getName());
+
+  private static final String MESHCA_URL_KEY = "meshCaUrl";
+  private static final String RPC_TIMEOUT_SECONDS_KEY = "rpcTimeoutSeconds";
+  private static final String GKECLUSTER_URL_KEY = "gkeClusterUrl";
+  private static final String CERT_VALIDITY_SECONDS_KEY = "certValiditySeconds";
+  private static final String RENEWAL_GRACE_PERIOD_SECONDS_KEY = "renewalGracePeriodSeconds";
+  private static final String KEY_ALGO_KEY = "keyAlgo";  // aka keyType
+  private static final String KEY_SIZE_KEY = "keySize";
+  private static final String SIGNATURE_ALGO_KEY = "signatureAlgo";
+  private static final String MAX_RETRY_ATTEMPTS_KEY = "maxRetryAttempts";
+  private static final String STS_URL_KEY = "stsUrl";
+  private static final String GKE_SA_JWT_LOCATION_KEY = "gkeSaJwtLocation";
+
+  static final String MESHCA_URL_DEFAULT = "meshca.googleapis.com";
+  static final long RPC_TIMEOUT_SECONDS_DEFAULT = 5L;
+  static final long CERT_VALIDITY_SECONDS_DEFAULT = 9L * 3600L; // 9 hours
+  static final long RENEWAL_GRACE_PERIOD_SECONDS_DEFAULT = 1L * 3600L; // 1 hour
+  static final String KEY_ALGO_DEFAULT = "RSA";  // aka keyType
+  static final int KEY_SIZE_DEFAULT = 2048;
+  static final String SIGNATURE_ALGO_DEFAULT = "SHA256withRSA";
+  static final int MAX_RETRY_ATTEMPTS_DEFAULT = 3;
+  static final String STS_URL_DEFAULT = "https://securetoken.googleapis.com/v1/identitybindingtoken";
+
+  private static final Pattern CLUSTER_URL_PATTERN = Pattern
+      .compile(".*/projects/(.*)/locations/(.*)/clusters/.*");
+
+  private static final String TRUST_DOMAIN_SUFFIX = ".svc.id.goog";
+  private static final String AUDIENCE_PREFIX = "identitynamespace:";
+  static final String MESH_CA_NAME = "meshCA";
+
+  static {
+    try {
+      CertificateProviderRegistry.getInstance()
+          .register(
+              new MeshCaCertificateProviderProvider(
+                  StsCredentials.Factory.getInstance(),
+                  MeshCaCertificateProvider.MeshCaChannelFactory.getInstance(),
+                  new ExponentialBackoffPolicy.Provider(),
+                  MeshCaCertificateProvider.Factory.getInstance())
+        );
+    } catch (Throwable t) {
```

Comment:
What's the possible cause of throwable here?

Reply:
Good point. Also catching and not re-throwing was bad too. Removed the catch clause.